session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift

```diff
@@ -0,0 +1,231 @@
+/// Like original algorithm, our implementation uses 6 passes over both arrays to determine diff
+///
+/// Ref.:
+/// - _"A technique for isolating differences between files"_ Paul Heckel (1978) - https://dl.acm.org/citation.cfm?id=359467
+///
+/// - Parameters:
+///   - oldArray: original array
+///   - newArray: new array
+/// - Returns: `Diff` describing changes from `oldArray` to `newArray`.
+internal func computeDiff<E: Diffable>(oldArray: [E], newArray: [E]) throws -> Diff<E> {
+    var table: [DiffableID: Symbol] = [:]
+    var oa: [Entry] = [] // old array entries
+    var na: [Entry] = [] // new array entries
+
+    // 1st pass
+    // Read `newArray` and store info on each element in symbols `table`:
+    for element in newArray {
+        table[element.id] = Symbol(inOld: false, inNew: true, indexInOld: nil)
+        na.append(.reference(element.id))
+    }
+
+    // 2nd pass
+    // Read `oldArray` and store info on each element in symbols `table`. If certain element already
+    // exists, update its information (otherwise create new entry):
+    for (index, element) in oldArray.enumerated() {
+        if table[element.id] == nil {
+            table[element.id] = Symbol(inOld: true, inNew: false, indexInOld: index)
+        } else {
+            table[element.id]?.inOld = true
+            table[element.id]?.indexInOld = index
+        }
+        oa.append(.reference(element.id))
+    }
+
+    // 3rd pass
+    // Uses "Observation 1":
+    // > If a line occurs only once in each file, then it must be the same line, although it may have been moved.
+    // > We use this observation to locate unaltered lines that we subsequently exclude from further treatment.
+    for (index, entry) in na.enumerated() {
+        if case let .reference(id) = entry {
+            guard let symbol = table[id] else {
+                throw DiffError()
+            }
+            if symbol.inOld && symbol.inNew {
+                guard let indexInOld = symbol.indexInOld else {
+                    throw DiffError()
+                }
+                na[index] = .index(indexInOld)
+                oa[indexInOld] = .index(index)
+            }
+        }
+    }
+
+    // 4th pass
+    // > If a line has been found to be unaltered, and the lines immediately adjacent to it in both files are identical,
+    // > then these lines must be the same line. This information can be used to find blocks of unchanged lines.
+    if na.count > 1 {
+        for i in (1..<(na.count - 1)) {
+            if case let .index(j) = na[i], j + 1 < oa.count {
+                if case let .reference(id1) = na[i + 1], case let .reference(id2) = oa[j + 1], id1 == id2 {
+                    na[i + 1] = .index(j + 1)
+                    oa[j + 1] = .index(i + 1)
+                }
+            }
+        }
+    }
+
+    // 5th pass
+    // Similar to 4th pass, except it processes entries in descending order.
+    if na.count > 1 {
+        for i in (1..<(na.count - 1)).reversed() {
```

Comment:
here you are missing the last element of na. `na[count-1]`. Is this intended ?

Reply:
Indeed, good catch - it should loop through `(1..<na.count)` range - fixed 👍👌. Similar to https://github.com/DataDog/dd-sdk-ios/pull/1005#discussion_r995935265, I couldn't figure out if it causes any problem even after enhancing fuzzy tests.